File: ruby-watir-tests/spec/pages/login_page.rb
# Page Object Model - Login Page
# Encapsulates all interactions with the login page
class LoginPage
  attr_reader :browser

  def initialize(browser)
    @browser = browser
  end

  # ---- Page URL ----
  def visit
    browser.goto("#{BASE_URL}/")
    self
  end

  # ---- Element Accessors ----
  def username_field
    browser.text_field(id: "username")
  end

  def password_field
    browser.text_field(id: "password")
  end

  def signin_button
    browser.button(id: "signin-button")
  end

  def message_element
    browser.element(id: "message")
  end

  def page_title
    browser.h1
  end

  # ---- Actions ----
  def enter_username(username)
    username_field.set(username)
    self
  end

  def enter_password(password)
    password_field.set(password)
    self
  end

  def click_signin
    signin_button.click
    self
  end

  def login(username, password)
    enter_username(username)
    enter_password(password)
    click_signin
    self
  end

  # ---- Assertions / Queries ----
  def displayed?
    username_field.present? && password_field.present? && signin_button.present?
  end

  def message_text
    # Wait for message to appear
    browser.wait_until(timeout: 5) { message_element.present? }
    message_element.text
  end

  def error_message?
    message_element.present? && message_element.class_name.include?("error-message")
  end

  def success_message?
    message_element.present? && message_element.class_name.include?("success-message")
  end

  def title_text
    page_title.text
  end
end

File: playwright-tests/tests/pages/login-page.ts
import { Page, Locator, expect } from "@playwright/test";

/**
 * Page Object Model - Login Page
 * Encapsulates all interactions with the login page
 */
export class LoginPage {
  readonly page: Page;
  readonly usernameField: Locator;
  readonly passwordField: Locator;
  readonly signinButton: Locator;
  readonly messageElement: Locator;
  readonly pageTitle: Locator;

  constructor(page: Page) {
    this.page = page;
    this.usernameField = page.locator("#username");
    this.passwordField = page.locator("#password");
    this.signinButton = page.locator("#signin-button");
    this.messageElement = page.locator("#message");
    this.pageTitle = page.locator("h1");
  }

  // ---- Page URL ----
  async visit() {
    await this.page.goto("/");
  }

  // ---- Actions ----
  async enterUsername(username: string) {
    await this.usernameField.fill(username);
  }

  async enterPassword(password: string) {
    await this.passwordField.fill(password);
  }

  async clickSignin() {
    await this.signinButton.click();
  }

  async login(username: string, password: string) {
    await this.enterUsername(username);
    await this.enterPassword(password);
    await this.clickSignin();
  }

  // ---- Assertions / Queries ----
  async isDisplayed() {
    await expect(this.usernameField).toBeVisible();
    await expect(this.passwordField).toBeVisible();
    await expect(this.signinButton).toBeVisible();
  }

  async getMessageText(): Promise<string> {
    await expect(this.messageElement).toBeVisible({ timeout: 5000 });
    return await this.messageElement.innerText();
  }

  async isErrorMessage(): Promise<boolean> {
    const className = await this.messageElement.getAttribute("class");
    return className?.includes("error-message") ?? false;
  }

  async isSuccessMessage(): Promise<boolean> {
    const className = await this.messageElement.getAttribute("class");
    return className?.includes("success-message") ?? false;
  }

  async getTitleText(): Promise<string> {
    return await this.pageTitle.innerText();
  }
}
